rubix-hadoop2/src/main/java/com/qubole/rubix/hadoop2/Hadoop2ClusterManager.java

```diff
@@ -0,0 +1,212 @@
+    static String ADDRESS = "yarn.resourcemanager.webapp.address";
+
+    @Override
+    public void initialize(Configuration conf)
+    {
+        super.initialize(conf);
+        log.debug("Initializing: Hadoop2ClusterManager");
+        this.address = conf.get(ADDRESS, address);
+        this.serverAddress = address.substring(0, address.indexOf(":"));
+        this.serverPort = Integer.parseInt(address.substring(address.indexOf(":") + 1));
+        nodesSupplier = Suppliers.memoizeWithExpiration(new Supplier<List<String>>()
+        {
+
+            @Override
+            public List<String> get()
+            {
+                if (!isMaster) {
+                    // First time all nodes start assuming themselves as master and down the line figure out their role
+                    // Next time onwards, only master will be fetching the list of nodes
+                    return ImmutableList.of();
+                }
+                try {
+                    StringBuffer response = new StringBuffer();
+                    URL obj = getNodeURL();
+                    HttpURLConnection httpcon = (HttpURLConnection) obj.openConnection();
+                    httpcon.setRequestMethod("GET");
+                    int responseCode = httpcon.getResponseCode();
+                    log.info("Sending 'GET' request to URL: " + obj.toString());
+                    if (responseCode == HttpURLConnection.HTTP_OK) {
+
+                        BufferedReader in = new BufferedReader(new InputStreamReader(httpcon.getInputStream()));
+                        String inputLine;
+                        while ((inputLine = in.readLine()) != null) {
+                            response.append(inputLine);
+                        }
+                        in.close();
+                        httpcon.disconnect();
+                    }
+                    else {
+                        log.info("/ws/v1/cluster/nodes failed due to " + responseCode + ". Setting this node as worker.");
+                        isMaster = false;
+                        httpcon.disconnect();
+                        return ImmutableList.of();
+                    }
+                    Gson gson = new Gson();
+                    Type type = new TypeToken<Nodes>() {}.getType();
+                    Nodes nodes = gson.fromJson(response.toString(), type);
+                    List<Elements> allNodes = nodes.getNodes().getNode();
+                    List<Elements> unhealthyNodes = new ArrayList<Elements>();
+                    for (Elements node : allNodes) {
+                        String state = node.getState();
+                        String nodeHostName = node.getNodeHostName();
+                        log.debug("Hostname: " + nodeHostName + "State: " + state);
+                        if (!state.equalsIgnoreCase("Running") && !state.equalsIgnoreCase("New") && !state.equalsIgnoreCase("Rebooted")) {
+                            unhealthyNodes.add(node);
```

Comment:
skip this logic of creating unhealthuNodes list.

Directly add to Set<String> hosts here when node is in one of these valid states